File: app/src/main/java/com/huanchengfly/tieba/post/activities/HistoryActivity.java
package com.huanchengfly.tieba.post.activities;

import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.ListView;
import android.widget.Toast;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.widget.Toolbar;

import com.google.gson.JsonSyntaxException;
import com.huanchengfly.tieba.post.ui.theme.utils.ThemeUtils;
import com.huanchengfly.tieba.post.R;
import com.huanchengfly.tieba.post.adapters.HistoryListAdapter;
import com.huanchengfly.tieba.post.models.ThreadHistoryInfoBean;
import com.huanchengfly.tieba.post.models.database.History;
import com.huanchengfly.tieba.post.utils.DisplayUtil;
import com.huanchengfly.tieba.post.utils.HistoryHelper;
import com.huanchengfly.tieba.post.utils.NavigationHelper;
import com.huanchengfly.tieba.post.utils.ThemeUtil;
import com.huanchengfly.tieba.post.utils.GsonUtil;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class HistoryActivity extends BaseActivity {
    private HistoryHelper helper;
    private NavigationHelper navigationHelper;
    private ListView mListView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_history);
        ThemeUtil.setTranslucentThemeBackground(findViewById(R.id.background));
        navigationHelper = NavigationHelper.newInstance(this);
        mListView = (ListView) findViewById(R.id.history_list);
        mListView.setDivider(new ColorDrawable(ThemeUtils.getColorByAttr(this, R.attr.colorDivider)));
        mListView.setDividerHeight(DisplayUtil.dp2px(this, 1));
        mListView.setOnItemClickListener((parent, view, position, id) -> {
            HistoryListAdapter adapter = (HistoryListAdapter) mListView.getAdapter();
            History dataBean = (History) adapter.getItem(position);
            switch (dataBean.getType()) {
                case HistoryHelper.TYPE_URL:
                    navigationHelper.navigationByData(NavigationHelper.ACTION_URL, dataBean.getData());
                    break;
                case HistoryHelper.TYPE_BA:
                    navigationHelper.navigationByData(NavigationHelper.ACTION_FORUM, dataBean.getData());
                    break;
                case HistoryHelper.TYPE_THREAD:
                    if (dataBean.getExtras() != null) {
                        try {
                            ThreadHistoryInfoBean historyInfoBean = GsonUtil.getGson().fromJson(dataBean.getExtras(), ThreadHistoryInfoBean.class);
                            if (historyInfoBean == null) {
                                break;
                            }
                            Map<String, String> map = new HashMap<>();
                            map.put("tid", dataBean.getData());
                            map.put("pid", historyInfoBean.getPid());
                            map.put("from", ThreadActivity.FROM_HISTORY);
                            map.put("seeLz", historyInfoBean.isSeeLz() ? "1" : "0");
                            navigationHelper.navigationByData(NavigationHelper.ACTION_THREAD, map);
                            break;
                        } catch (JsonSyntaxException ignored) {
                        }
                    }
                    Map<String, String> map = new HashMap<>();
                    map.put("tid", dataBean.getData());
                    navigationHelper.navigationByData(NavigationHelper.ACTION_THREAD, map);
                    break;
            }
        });
        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
            actionBar.setTitle(R.string.title_history);
        }
        helper = new HistoryHelper(this);
        refreshData();
    }

    private void refreshData() {
        List<History> array = helper.getAll();
        mListView.setAdapter(new HistoryListAdapter(this, array));
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_history_toolbar, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
            case R.id.menu_delete:
                helper.delete();
                Toast.makeText(this, R.string.toast_delete_success, Toast.LENGTH_SHORT).show();
                refreshData();
                break;
            default:
                break;
        }
        return super.onOptionsItemSelected(item);
    }
}

File: app/src/main/java/com/huanchengfly/tieba/post/activities/HistoryActivity.kt
package com.huanchengfly.tieba.post.activities

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.Toast
import androidx.appcompat.widget.Toolbar
import androidx.viewpager.widget.ViewPager
import butterknife.BindView
import com.google.android.material.appbar.CollapsingToolbarLayout
import com.google.android.material.tabs.TabLayout
import com.huanchengfly.tieba.post.R
import com.huanchengfly.tieba.post.adapters.FragmentTabViewPagerAdapter
import com.huanchengfly.tieba.post.fragments.HistoryFragment
import com.huanchengfly.tieba.post.interfaces.Refreshable
import com.huanchengfly.tieba.post.utils.HistoryUtil
import com.huanchengfly.tieba.post.utils.ThemeUtil

class HistoryActivity : BaseActivity() {
    @BindView(R.id.view_pager)
    lateinit var viewPager: ViewPager

    @BindView(R.id.toolbar)
    lateinit var toolbar: Toolbar

    @BindView(R.id.collapsing_toolbar)
    lateinit var collapsingToolbar: CollapsingToolbarLayout

    @BindView(R.id.tab_layout)
    lateinit var tabLayout: TabLayout

    var loaded = false

    private val viewPagerAdapter by lazy { FragmentTabViewPagerAdapter(supportFragmentManager) }

    override fun getLayoutId(): Int = R.layout.activity_history

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ThemeUtil.setTranslucentThemeBackground(findViewById(R.id.background))
        toolbar.title = title
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        collapsingToolbar.title = title
        collapsingToolbar.isTitleEnabled = true
        viewPagerAdapter.addFragment(HistoryFragment.newInstance(HistoryUtil.TYPE_THREAD), getString(R.string.title_history_thread))
        viewPagerAdapter.addFragment(HistoryFragment.newInstance(HistoryUtil.TYPE_FORUM), getString(R.string.title_history_forum))
        viewPager.adapter = viewPagerAdapter
        tabLayout.setupWithViewPager(viewPager)
    }

    override fun onResume() {
        super.onResume()
        if (loaded) refreshData()
        loaded = true
    }

    private fun refreshData() {
        viewPagerAdapter.fragments.forEach {
            if (it is Refreshable) {
                it.onRefresh()
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_history_toolbar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.menu_delete -> {
                HistoryUtil.deleteAll()
                Toast.makeText(this, R.string.toast_delete_success, Toast.LENGTH_SHORT).show()
                refreshData()
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}